Intern/rayx-ui/src/Triangulation/Triangulate.cpp

```diff
@@ -2,128 +2,423 @@
 
 #include <glm/gtc/matrix_transform.hpp>
 #include <glm/gtx/string_cast.hpp>
+#include <map>
 #include <memory>
+#include <queue>
+#include <set>
 
 #include "Colors.h"
 #include "Debug/Debug.h"
 #include "GeometryUtils.h"
 #include "Shader/Constants.h"
 #include "Triangulation/TraceTriangulation.h"
 
+struct Point2D {
+    double x;
+    double y;
+
+    int operator<(Point2D other) const { return y < other.y || (y == other.y && x > other.x); }
+    int operator>(Point2D other) const { return y > other.y || (y == other.y && x < other.x); }
+
+    int operator==(Point2D other) const { return x == other.x && y == other.y; }
+};
+
+// A simple polygon is a list of points in counter-clockwise order
+typedef std::vector<uint32_t> PolygonSimple;
+
+// A complex polygon is a list of simple polygons
+// holes are represented by polygons in clockwise order
+typedef std::vector<PolygonSimple> PolygonComplex;
+
+typedef std::vector<uint32_t> Triangles;
+
+enum VertexType {
+    Start,
+    Split,
+    End,
+    Merge,
+    Regular,
+};
+
+struct Vertex {
+    uint32_t index;
+    VertexType type;
+};
+
+struct EdgeEntry {
+    Vertex origin;
+    uint32_t prev;
+    uint32_t next;
+};
+
+double absoluteAngle(Point2D p1, Point2D p2) { return atan2(p2.x - p1.x, p2.y - p1.y); }
+
+VertexType toVertexType(Point2D prev, Point2D current, Point2D next) {
+    double angle = atan2(next.x - current.x, next.y - current.y) - atan2(prev.x - current.x, prev.y - current.y);
+    if (angle < 0) {
+        angle += M_PI * 2;
+    }
+    if (prev < current && next < current) {
+        return angle < M_PI ? Start : Split;
+    } else if (prev > current && next > current) {
+        return angle < M_PI ? End : Merge;
+    } else {
+        return Regular;
+    }
+}
+
+enum ChainType {
+    Left,
+    Right,
+};
+
+// translate a possibly negative index to a positive one
+int translateIndex(int index, int size) { return (index + size) % size; }
+
+struct EdgeList {
+    std::vector<EdgeEntry> edges;
+    std::vector<Point2D> points;
+
+    Point2D getOrigin(uint32_t edge) { return points[edges[edge].origin.index]; }
+
+    uint32_t getOriginIndex(uint32_t edge) { return edges[edge].origin.index; }
+
+    VertexType getOriginType(uint32_t edge) { return edges[edge].origin.type; }
+
+    uint32_t getPrev(uint32_t edge) { return edges[edge].prev; }
+
+    uint32_t getNext(uint32_t edge) { return edges[edge].next; }
+
+    double xAtY(uint32_t edge, double y) {
+        Point2D p1 = getOrigin(edge);
+        Point2D p2 = getOrigin(getNext(edge));
+
+        if (p1.y == p1.y) {
+            return (p1.x + p2.x) / 2;
+        }
+
+        return p1.x + (y - p1.y) * (p2.x - p1.x) / (p2.y - p1.y);
+    }
+
+    EdgeList(PolygonComplex poly, std::vector<TextureVertex>& verts) {
+        for (auto& vert : verts) {
+            points.push_back({-vert.pos.x, vert.pos.z});
+        }
+        uint32_t startI = 0;
```

Comment:
Something like `edges.reserve(verts.size());` can be a decent speed up if the vector would otherwise resize itself regularly. It's not necessarily important right now (due to low vertex counts), but it might future-proof this a bit more.